cpp/src/tsne/bh_kernels.cuh

```diff
@@ -635,17 +655,37 @@ __global__ void attractive_kernel_bh(
   const float *restrict VAL, const int *restrict COL, const int *restrict ROW,
   const float *restrict Y1, const float *restrict Y2,
   const float *restrict norm, float *restrict attract1,
-  float *restrict attract2, const int NNZ) {
+  float *restrict attract2, const int NNZ,
+  int *restrict flag_unstable_computation) {
   const int index = (blockIdx.x * blockDim.x) + threadIdx.x;
   if (index >= NNZ) return;
   const int i = ROW[index];
   const int j = COL[index];
 
   // TODO: Calculate Kullback-Leibler divergence
   // TODO: Convert attractive forces to CSR format
-  const float PQ = __fdividef(
-    VAL[index],
-    norm[i] + 1.0f + norm[j] - 2.0f * (Y1[i] * Y1[j] + Y2[i] * Y2[j]));  // P*Q
+  // Try single precision compute first
+  float denominator = __fmaf_rn(-2.0f, (Y1[i] * Y1[j]), norm[i] + 1.0f) +
+                      __fmaf_rn(-2.0f, (Y2[i] * Y2[j]), norm[j]);
+
+  if (__builtin_expect(denominator == 0, false)) {
+    double _Y1 = static_cast<double>(Y1[i] * Y1[j]);
+    double _Y2 = static_cast<double>(Y2[i] * Y2[j]);
+    double dbl_denominator =
+      __fma_rn(-2.0f, _Y1, norm[i] + 1.0f) + __fma_rn(-2.0f, _Y2, norm[j]);
```

Comment:
It might be helpful to open a Github issue for further investigating this detail and reference it here for future eyes.

Reply:
@zbjornson Can you link the PR you mention in your other comment?